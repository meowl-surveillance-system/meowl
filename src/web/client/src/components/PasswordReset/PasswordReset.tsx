import React, {
  Component,
  FormEvent,
  ChangeEvent,
  ComponentState,
} from "react";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Container from "@material-ui/core/Container";

interface Props {
  match: any;
  history: any;
}
interface State {
  password: string;
  successMessage: string;
  failMessage: string;
}

export default class PasswordReset extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      password: "",
      successMessage: "",
      failMessage: "",
    };
  }

  async componentDidMount() {
    try {
      const requestOptions = {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          resetToken: this.props.match.params.token,
        }),
      };
      const res = await fetch("/auth/verifyToken", requestOptions);
      if (res.status === 400) {
        this.props.history.push("/");
      }
    } catch (e) {
      console.error(e);
    }
  }

  onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    try {
      event.preventDefault();
      const requestOptions = {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          password: this.state.password,
          resetToken: this.props.match.params.token,
        }),
      };
      const res = await fetch("/auth/submitPasswordReset", requestOptions);
      const msg = await res.json();
      if (msg === "Successfully updated password") {
        this.setState({ successMessage: msg, failMessage: "" });
      } else {
        this.setState({ failMessage: msg, successMessage: "" });
      }
    } catch (e) {
      console.error(e);
    }
  };

  onChange = async (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    this.setState({
      [event.target.name]: event.target.value,
    } as ComponentState);
  };

  render() {
    return (
      <Container component="main" maxWidth="xs">
        <Grid
          container
          spacing={2}
          direction="column"
          alignItems="center"
          justify="center"
          style={{ minHeight: "100vh" }}
        >
          {this.state.failMessage && (
            <Typography style={{ color: "red" }}>
              {this.state.failMessage}
            </Typography>
          )}
          {this.state.successMessage && (
            <Typography style={{ color: "green" }}>
              {this.state.successMessage}
            </Typography>
          )}
          <Typography component="h1" variant="h5">
            Password Reset Form
          </Typography>
          <form noValidate onSubmit={this.onSubmit}>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="password"
              label="Password"
              type="password"
              onChange={this.onChange}
            />
            <Button fullWidth type="submit" variant="contained" color="primary">
              Submit
            </Button>
          </form>
        </Grid>
      </Container>
    );
  }
}
