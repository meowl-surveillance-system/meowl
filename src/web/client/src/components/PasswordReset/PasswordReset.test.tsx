import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import PasswordReset from "./PasswordReset";

describe("PasswordReset component", () => {
  let passwordResetWrapper: ShallowWrapper<{}, {}, PasswordReset>;
  let matchMock: object;
  let historyMock: object;
  let eventMock: any;
  let preventDefaultMock: jest.Mock;
  let mockSuccessResponse: string;
  let mockJsonPromise: Promise<string>;
  let mockFetchPromise: Promise<object>;
  beforeEach(() => {
    mockSuccessResponse = "Successfully updated password";
    mockJsonPromise = Promise.resolve(mockSuccessResponse);
    mockFetchPromise = Promise.resolve({ json: () => mockJsonPromise });
    jest
      .spyOn(window, "fetch")
      .mockImplementation(() => mockFetchPromise as Promise<Response>);
    matchMock = {
      params: {
        token: "muda",
      },
    };
    historyMock = {
      push: jest.fn(),
    };
    preventDefaultMock = jest.fn();
    eventMock = {
      preventDefault: preventDefaultMock,
      target: {
        name: "password",
        value: "beeetch",
      },
    };
    passwordResetWrapper = shallow(
      <PasswordReset match={matchMock} history={historyMock} />,
    );
  });

  it("renders PasswordReset component", () => {
    const { container } = render(
      <Router>
        <PasswordReset match={matchMock} history={historyMock} />
      </Router>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should set successMessage on successful resolve", async () => {
    await passwordResetWrapper.instance().onSubmit(eventMock);
    expect(passwordResetWrapper.instance().state.successMessage).toEqual(
      mockSuccessResponse,
    );
  });
  it("should not set failMessage on promise reject", async () => {
    await passwordResetWrapper.instance().onSubmit(eventMock);
    expect(passwordResetWrapper.instance().state.failMessage).toEqual("");
  });
  it("should set password field when onChange is called", () => {
    passwordResetWrapper.instance().onChange(eventMock);
    expect(passwordResetWrapper.instance().state.password).toEqual(
      eventMock.target.value,
    );
  });
});
