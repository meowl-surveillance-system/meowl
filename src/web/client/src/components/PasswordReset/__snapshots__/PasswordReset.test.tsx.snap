// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordReset component renders PasswordReset component 1`] = `
<main
  class="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    style="min-height: 100vh;"
  >
    
    
    <h1
      class="MuiTypography-root MuiTypography-h5"
    >
      Password Reset Form
    </h1>
    <form
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          Password
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="password"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-200 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-202"
            >
              <span>
                Password
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</main>
`;
